SONAR-6569 move Quality Gate events to CE + persist CE events to DB

sns-seb opened the pull request.

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java:
```diff
@@ -0,0 +1,133 @@
+    String alertText = currentStatus.getAlertText();
+    MetricLevel alertLevel = MetricLevel.valueOf(currentStatus.getAlertStatus());
+    if (pastStatus.isPresent()) {
+      MetricLevel pastAlertLevel = MetricLevel.valueOf(pastStatus.get().getAlertStatus());
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Possible null pointer dereference in org.sonar.server.computation.step.QualityGateEventsStep.checkQualityGateStatusChange(Component, BatchReport$Measure, Optional) due to return value of called method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java:
```diff
@@ -0,0 +1,133 @@
+    }
+  }
+
+  // protected void notifyUsers(Component project, String alertName, String alertText, MetricLevel alertLevel, boolean isNewAlert) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

julienlancelot replied:
Why leave this commented code ?

sns-seb replied:
I forsee it will be done in the same step, hence the code left commented out for now

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java:
```diff
@@ -0,0 +1,133 @@
+      // There were no defined alerts before, so this one is a new one
+      String alertName = buildName(currentStatus);
+      createEvent(project, alertName, alertText);
+      // notifyUsers(project, alertName, alertText, alertLevel, true);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java:
```diff
@@ -0,0 +1,133 @@
+  // .setFieldValue("alertName", alertName)
+  // .setFieldValue("alertText", alertText)
+  // .setFieldValue("alertLevel", alertLevel.toString())
+  // .setFieldValue("isNewAlert", Boolean.toString(isNewAlert));
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

SonarTech commented:
SonarQube analysis reported 8 new issues:
* 3 critical
* 5 major

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  private String buildName(MeasureDto currentStatus) {
+    return MetricLevel.valueOf(currentStatus.getAlertStatus()).getColorName();
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Possible null pointer dereference in org.sonar.server.computation.step.QualityGateEventsStep.buildName(MeasureDto) due to return value of called method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java:
```diff
@@ -0,0 +1,133 @@
+        String alertName = buildName(pastStatus.get(), currentStatus);
+        createEvent(project, alertName, alertText);
+        // There was already a Orange/Red alert, so this is no new alert: it has just changed
+        // boolean isNewAlert = pastAlertLevel == MetricLevel.OK;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

julienlancelot replied:
These 2 lines should be removed

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistEventsStep.java:
```diff
@@ -143,4 +213,13 @@ private String convertCategory(Constants.EventCategory category) {
+
+    @Override
+    public Integer apply(EventDto input) {
+      return Objects.hashCode(input.getComponentUuid(), input.getSnapshotId(), input.getDate(), input.getCategory(), input.getName(), input.getDescription(), input.getData());
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) input must be non-null but is marked as nullable [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)

SonarTech commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistEventsStep.java:
```diff
@@ -31,111 +41,171 @@
     try {
-      new EventVisitor(session, reportReader.readMetadata().getAnalysisDate()).visit(treeRootHolder.getRoot());
+      final long analysisDate = reportReader.readMetadata().getAnalysisDate();
+      new DepthTraversalTypeAwareVisitor(Component.Type.FILE, DepthTraversalTypeAwareVisitor.Order.PRE_ORDER) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this anonymous class number of lines from 21 to at most 20, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1188)

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java:
```diff
@@ -0,0 +1,133 @@
+        createEvent(project, alertName, alertText);
+        // There was already a Orange/Red alert, so this is no new alert: it has just changed
+        // boolean isNewAlert = pastAlertLevel == MetricLevel.OK;
+        // notifyUsers(project, alertName, alertText, alertLevel, isNewAlert);
```
Is notification will be done in this step or later ?

sns-seb replied:
I forsee it will be done in the same step, hence the code left commented out for now

julienlancelot replied:
So you should add a TODO in order to not forget it. Otherwise it's hard to understand why there's dead code here

julienlancelot commented on server/sonar-server/src/test/java/org/sonar/server/computation/step/QualityGateEventsStepTest.java:
```diff
@@ -0,0 +1,167 @@
+  private ArgumentCaptor<Event> eventArgumentCaptor = ArgumentCaptor.forClass(Event.class);
+
+  private EventRepository eventRepository = mock(EventRepository.class);
+  private MeasureRepository measureRepository = mock(MeasureRepository.class);
```
Dont' you think that a fake implementation of repositories should be created for test ? It would remove the needs to mock it

sns-seb replied:
I make sure there is no unwanted calls to the repositories in tests, so mocks are more appropriate

julienlancelot replied:
I understand, but using mock could make refactoring harder. But ok for now

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistEventsStep.java:
```diff
@@ -31,111 +41,171 @@
+
+        @Override
+        public void visitDirectory(Component directory) {
+          processComponent(directory, session, analysisDate);
```
AFAIK, no events are generated on directories and files

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistEventsStep.java:
```diff
@@ -31,111 +41,171 @@
+
+        @Override
+        public void visitFile(Component file) {
+          processComponent(file, session, analysisDate);
```
AFAIK, no events are generated on directories and files

sns-seb replied:
well, we could make that optimization, it would work since, for now, we only create events on projects (not even sure on modules).

But design allows defining events on any component, if we want this step to comply with the CE design, we should saves events on any Component. The extra cost can be estimated as little so I'd rather save on any Component until proven otherwise.

julienlancelot replied:
Ok

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistEventsStep.java:
```diff
@@ -31,111 +41,171 @@
+    if (!difference.isEmpty()) {
+      for (Integer hash : difference) {
+        Loggers.get(PersistEventsStep.class).error(
+          String.format("Unmatching EventDto for Component %s batch='%s' ce='%s'",
```
I'm not sure we should keep this code in production

sns-seb commented:
PR updated:
* fixed quality flaws reported by SQ
* forced me to support invalid values in measure => code logic has a been a little extended in `QualityGateEventsStep`
* rebased on master

julienlancelot commented:
LGTM !
